### PR TITLE
Reconnectable and Shared Mono<RSocket> feature

### PR DESCRIPTION
**Pull request by OlegDokuka**
This PR enables a reconnectable RSocket Mono feature which allows to:

1) Have shared `Mono` instance without extra effort (extra effort == `mono.cache()`)
2) Have allocation free blocking (`ReconnectMono#block`) as in the case of `MonoProcessor`
3) Have a `Mono` which reestablishes connection once the previous has been expired

```java
Mono<RSocket> sharedRSocketMono =          
  RSocketFactory                           
               .connect()                  
               .singleSubscriberRequester()
               .reconnect()                
               .transport(transport)       
               .start();                   
                                           
 RSocket r1 = sharedRSocketMono.block();   // subscribes first so the connection establishement is initiated 
 RSocket r2 = sharedRSocketMono.block();   // attached to the previous process
                                           
 assert r1 == r2;    // same reference                       
                                           
 r1.dispose(); // disposes instance so it is not valid anymore                              
                                           
 assert r2.isDisposed()                    
                                           
 RSocket r3 = sharedRSocketMono.block();   // reistablish a new connection
 RSocket r4 = sharedRSocketMono.block();   // attached to the previous process
                                           
                                           
 assert r1 != r3; // new instance is give (ref check is not valid)                         
 assert r4 == r3; // r3 and r4 are ref equal 
                 
 assert r2.isDisposed(); // previous is still disposed
 assert !r4.isDisposed(); // a new one is healthy 	
                        
```

Signed-off-by: Oleh Dokuka <shadowgun@i.ua>

### CODE REVIEW

**Comment by OlegDokuka** [comment]
cc @rstoyanchev @simonbasle for your review folks

File: rsocket-core/src/main/java/io/rsocket/Invalidate.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.rsocket;
+
+interface Invalidate {
+
+  void expire();
```
Comment (simonbasle) [inline rsocket-core/src/main/java/io/rsocket/Invalidate.java]: on caches the terminology is typically `invalidate()`, so I'd go with that. Maybe rename the class to `Invalidatable` for drawing a parallel with `Disposable`/`Closeable`/etc... (even though that's not a word 😆)?

File: rsocket-core/src/main/java/io/rsocket/RSocketFactory.java [hunk 2]
```diff
@@ -392,6 +523,17 @@ public ClientRSocketFactory frameDecoder(PayloadDecoder payloadDecoder) {
                                 .sendOne(setupFrame)
                                 .thenReturn(wrappedRSocketRequester);
                           });
+                })
+            .as(
+                source -> {
+                  if (reconnectEnabled) {
+                    return new ReconnectMono<>(
+                        source.retryWhen(whenFactory),
```
[the remark this answers is not in the document]
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/RSocketFactory.java]: whenFactort is an argument for the retryWhen. How should one apply retry logic then?

File: rsocket-core/src/main/java/io/rsocket/Invalidate.java (hunk 1 above)
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/Invalidate.java]: fixed

File: rsocket-core/src/main/java/io/rsocket/RSocketFactory.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonbasle) [reply rsocket-core/src/main/java/io/rsocket/RSocketFactory.java]:
```java
return new ReconnectMono<>(
    (whenFactory == FAIL_WHEN_FACTORY) ? source : source.retryWhen(whenFactory),
//...
```

[the remark this answers is not in the document]
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/RSocketFactory.java]: Gotcha

[the remark this answers is not in the document]
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/RSocketFactory.java]: fixed

**Comment by rstoyanchev** [comment]
Looks great, exposed through `ClientRSocketFactory` as opposed to the other alternatives considered. 

I'm wondering about using the new [Retry](https://projectreactor.io/docs/core/release/api/reactor/util/retry/Retry.html) from reactor-core which has some very nice built-in options.

For `reconnect()`, is it too simplistic maybe? Note that since the new `Retry` exposes a number of strategies it's much easier to configure. Perhaps we don't even need a no-arg `reconnect()` or do we think there is a good default reconnect strategy?

**Comment by OlegDokuka** [comment]
@rstoyanchev no-args `reconnect()` is basically fail fast version which does nothing if upstream fails. 
I prepared that one if a user does not want to retry at all but wants to have a reconnectable or shared rsocket. 

Do you think such an extension (with no retry) is unnecessary? 
> Note that since the new Retry exposes a 

Retry is now a convenient way to do retry, so I'm going to replace function with that one shortly

File: rsocket-core/src/main/java/io/rsocket/RSocketFactory.java [hunk 3]
```diff
@@ -230,6 +238,68 @@ public ClientRSocketFactory singleSubscriberRequester() {
...
+     *
+     *  RSocket r3 = sharedRSocketMono.block();
+     *  RSocket r4 = sharedRSocketMono.block();
+     *
+     *
+     *  assert r1 != r3;
+     *  assert r4 == r3;
+     *
+     * }</pre>
+     *
```
[the remark this answers is not in the document]
Comment (OlegDokuka) [reply rsocket-core/src/main/java/io/rsocket/RSocketFactory.java]: fixed

**Pull request merged**